File: leetcode/0034_在排序数组中查找元素的第一个和最后一个位置/solution.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
function searchRange(nums, target) {
  let i = 0, j = nums.length - 1
  let result = [-1, -1]
  let count = 0
  while (i <= j && count < 2) {
    console.log(nums[i], nums[j])
    if (result[0] === -1) {
      if (nums[i] === target) {
        result[0] = i
        count++
      } else {
        i++
      }
    }

    if (result[1] === -1) {
      if (nums[j] === target) {
        result[1] = j
        count++
      } else {
        j--
      }
    }
  }
  return result
}

const nums = [5,7,7,8,8,10]

console.log(searchRange(nums, 8))

File: leetcode/0034_在排序数组中查找元素的第一个和最后一个位置/solution.cpp
class Solution {
public:
    int binarySearch(vector<int>& nums, int target, bool flag) {
        int l = 0, r = nums.size() - 1;
        int ans = nums.size();
        while (l <= r) {
            int mid = (l + r) / 2;
            if (nums[mid] > target || (flag && nums[mid] >= target)) {
                r = mid - 1;
                ans = mid;
            } else {
                l = mid + 1;
            }
        }
        return ans;
    }

    vector<int> searchRange(vector<int>& nums, int target) {
        int leftIndex = binarySearch(nums, target, true);
        int rightIndex = binarySearch(nums, target, false) - 1;

        if (leftIndex <= rightIndex && rightIndex < nums.size() && nums[leftIndex] == target && nums[rightIndex] == target) {
            return vector<int> {leftIndex, rightIndex};
        }

        return vector<int> {-1, -1};
    }
};